8265491: Math Signum optimization for x86

mgkwill opened the pull request:
x86 Math.Signum() uses two floating point compares and a copy sign operation involving data movement to gpr and XMM.

We can optimize to one floating point compare and sign computation in XMM. We observe ~25% performance improvement with this optimization.

Base:
```
Benchmark                       Mode Cnt Score Error Units
Signum._1_signumFloatTest avgt 5 4.660 ? 0.040 ns/op
Signum._2_overheadFloat avgt 5 3.314 ? 0.023 ns/op
Signum._3_signumDoubleTest avgt 5 4.809 ? 0.043 ns/op
Signum._4_overheadDouble avgt 5 3.313 ? 0.015 ns/op
```
 
Optimized:
signum intrinsic patch
```
Benchmark                       Mode  Cnt  Score   Error  Units
Signum._1_signumFloatTest       avgt    5  3.769 ? 0.015  ns/op
Signum._2_overheadFloat         avgt    5  3.312 ? 0.025  ns/op
Signum._3_signumDoubleTest      avgt    5  3.765 ? 0.005  ns/op
Signum._4_overheadDouble        avgt    5  3.309 ? 0.010  ns/op
```

Signed-off-by: Marcus G K Williams <marcus.williams@intel.com>

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265491](https://bugs.openjdk.java.net/browse/JDK-8265491): Math Signum optimization for x86


### Reviewers
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - Committer)
 * [Jatin Bhateja](https://openjdk.java.net/census#jbhateja) (@jatin-bhateja - Committer)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3581/head:pull/3581` \
`$ git checkout pull/3581`

Update a local copy of the PR: \
`$ git checkout pull/3581` \
`$ git pull https://git.openjdk.java.net/jdk pull/3581/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3581`

View PR using the GUI difftool: \
`$ git pr show -t 3581`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3581.diff">https://git.openjdk.java.net/jdk/pull/3581.diff</a>

</details>

mgkwill commented:
/label hotspot-compiler

jatin-bhateja commented:
@mgkwill,  your newly added benchmark has 4 micro worklets, please publish the results for all of them.

DamonFool commented on test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java:
```diff
@@ -25,7 +25,7 @@
  * @test
  * @summary Test compiler intrinsics for signum
- * @requires os.arch=="aarch64"
+ * @requires os.arch=="aarch64" | os.arch=="x86" | os.arch=="amd64" | os.arch=="x86_64"
```
What about os.arch=="i386" ?

mgkwill replied:
@DamonFool, I wasn't sure about `os.arch=="i386"`. I'm still getting my head around os.arch and how it is used in these tests. 

I'll add it to the jtreg.

mgkwill commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5775,46 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
+  format %{ "signumF $dst, $dst" %}
+  ins_encode %{
+    Label exit;
+
```
Do you mean `instruct signumF_reg`? Please explain your reasoning for the proposed move.

jatin-bhateja replied:
I am referring to moving the instruction sequence into a macro assembly routine for better code sharing, one should be enough for both float and double type, please refer to following snippet for more detail.
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L5672

mgkwill replied:
Hi @jatin-bhateja your original request was quite unclear. The highlighting of the code was incorrect and the use of 'this' made it unclear what you were suggesting and you did not explain a reason for the suggestion. 

> Can you kindly move **this** into a macro assembly routine.

Seeing "better code sharing" as the reason it's unclear to me if the proposed refactor results in any meaningful benefit for the work involved. It's also unclear if there are any hard code structure or style issues that would necessitate the move.

The refactor you suggest calls for the instruction sequences here to be moved into a src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp function that is then called here. Inside that function there is a selection between float and double but otherwise stays the same except one is formatted .AD and one is .CPP/.HPP. 

If I am doing this refactor for `instruct signumF_reg` and `instruct signumD_reg` to cut down on code duplication, I see that as close to a wash for saving of code lines and effort involved. I also think this would make the code more complex to read.

However if there are strong opinions as to why this is necessary I'd be happy to hear them.

theRealAph replied:
I think it'd be easier to understand with a signum macro in MacroAssembler. Less code duplication is good, obviously. And it's potentially useful elsewhere. Win-win-win.
We've already factored out C2 instruction sequences for things like vabsnegd  into c2_MacroAssembler_x86.

mgkwill replied:
I wave the white flag. 

Thanks for the review and suggestions @theRealAph and @jatin-bhateja.

In an effort to eradicate the 5 lines of duplication and with an eye to potential usefulness in the future I've created a twice called `C2_MacroAssembler::signum_fp` using 36 lines of code and modification of c2_MacroAssembler_x86.cpp c2_MacroAssembler_x86.hpp.

mgkwill commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5775,46 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
+  ins_encode %{
+    Label exit;
+
+    __ ucomisd($dst$$XMMRegister, $zero$$XMMRegister);
```
Do you mean `instruct signumD_reg`? Please explain your reasoning for the proposed move.

DamonFool commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5775,46 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
 
+// --------------------------------- Signum ---------------------------
+
+instruct signumF_reg(regF dst, regF zero, regF one, rFlagsReg cr) %{
```
Do we need `predicate(UseSSE>=2);` here?

mgkwill replied:
It seems ucomiss is SSE, but ucomisd is SSE2. I will add a SSE predicate to `instruct signumF_reg` and an SSE2 predicate to `instruct signumD_reg`. Thanks for the suggestion.

mgkwill replied:
It looks like subss, subsd are also SSE, SSE2 respectively.

DamonFool commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5775,46 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
+    __ jcc(Assembler::equal, exit);
+    __ movflt($dst$$XMMRegister, $one$$XMMRegister);
+    __ jcc(Assembler::above, exit);
+    __ movflt($dst$$XMMRegister, $zero$$XMMRegister);
```
Is it possible to use just one instruction to assign -1 to $dst?

Maybe, you can try to follow negF_reg/negF_reg_reg.

mgkwill replied:
Good idea. Changed to use `xorps($dst$$XMMRegister, ExternalAddress(float_signflip()));` and `xorpd($dst$$XMMRegister, ExternalAddress(double_signflip()));`

DamonFool commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5775,46 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
+    Label exit;
+
+    __ ucomiss($dst$$XMMRegister, $zero$$XMMRegister);
+    __ jcc(Assembler::parity, exit);
```
How about checking equal first and then parity?

I think the unordered case is rare in real programs.

mgkwill replied:
+1

dholmes-ora commented on test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java:
```diff
@@ -25,7 +25,7 @@
  * @test
  * @summary Test compiler intrinsics for signum
- * @requires os.arch=="aarch64"
+ * @requires os.arch=="aarch64" | os.arch=="x86" | os.arch=="i386" | os.arch=="amd64" | os.arch=="x86_64"
```
Shouldn't this test run successfully on all platforms, even if they don't do anything in response to UseSignumIntrinsic? I can run it today on x86_64.

mgkwill replied:
Hi @dholmes-ora. I have not done negative testing on any platform using TestSignumIntrinsic.java. I'm not the original author, but what you say makes sense after some thought. I'll try an easy test, remove this line and see if presubmit tests pass.

mgkwill commented:
Updated benchmarks from patch updated from comments of @DamonFool:
```
Benchmark                       Mode  Cnt  Score   Error  Units
Signum._1_signumFloatTest       avgt    5  3.769 ? 0.015  ns/op
Signum._2_overheadFloat         avgt    5  3.312 ? 0.025  ns/op
Signum._3_signumDoubleTest      avgt    5  3.765 ? 0.005  ns/op
Signum._4_overheadDouble        avgt    5  3.309 ? 0.010  ns/op
```

Updated Summary.

DamonFool commented on src/hotspot/cpu/x86/vm_version_x86.cpp:
```diff
@@ -1700,6 +1700,9 @@ void VM_Version::get_processor_features() {
     }
   }
 #endif // !PRODUCT
+  if (FLAG_IS_DEFAULT(UseSignumIntrinsic) && (UseSSE >= 2)) {
```
`UseSSE >=2` looks a bit uncomfortable to me since signumF_reg only requires UseSSE>=1. 

How about something like this:
```
diff --git a/src/hotspot/cpu/x86/x86.ad b/src/hotspot/cpu/x86/x86.ad
index bbf5256c112..0a738704c4d 100644
--- a/src/hotspot/cpu/x86/x86.ad
+++ b/src/hotspot/cpu/x86/x86.ad
@@ -1598,6 +1598,16 @@ const bool Matcher::match_rule_supported(int opcode) {
         return false;
       }
       break;
+    case Op_SignumF:
+      if (UseSSE < 1) {
+        return false;
+      }
+      break;
+    case Op_SignumD:
+      if (UseSSE < 2) {
+        return false;
+      }
+      break;
 #endif // !LP64
   }
   return true;  // Match rules are supported by default.
diff --git a/src/hotspot/share/opto/library_call.cpp b/src/hotspot/share/opto/library_call.cpp
index 7cb7955c612..217ee39d709 100644
--- a/src/hotspot/share/opto/library_call.cpp
+++ b/src/hotspot/share/opto/library_call.cpp
@@ -1737,8 +1737,8 @@ bool LibraryCallKit::inline_math_native(vmIntrinsics::ID id) {
   case vmIntrinsics::_dpow:      return inline_math_pow();
   case vmIntrinsics::_dcopySign: return inline_double_math(id);
   case vmIntrinsics::_fcopySign: return inline_math(id);
-  case vmIntrinsics::_dsignum: return inline_double_math(id);
-  case vmIntrinsics::_fsignum: return inline_math(id);
+  case vmIntrinsics::_dsignum: return Matcher::match_rule_supported(Op_SignumD) ? inline_double_math(id) : false;
+  case vmIntrinsics::_fsignum: return Matcher::match_rule_supported(Op_SignumF) ? inline_math(id) : false;
 
    // These intrinsics are not yet correctly implemented
   case vmIntrinsics::_datan2:
```

In this way, we can enable UseSignumIntrinsic by default on x86.
Thanks.

mgkwill replied:
Thanks for the suggestion. This looks good as long as it doesn't affect aarch64.

mgkwill commented:
- Switched to 3 operand xorps/d on colleague suggestion to insure I don't introduce a hard to find register bug. 

- Implemented @DamonFool suggestion for Default Enable of UseSignumIntrinsic on x86. Slightly worried about `!has_match_rule(opcode)` for aarch64. But as long as it doesn't hit this I think it will work.

 src/hotspot/cpu/aarch64/aarch64.ad:
```
const bool Matcher::match_rule_supported(int opcode) {
  if (!has_match_rule(opcode))
    return false;
```

Also moved matcher checks in src/hotspot/cpu/x86/x86.ad outside of 
```
#ifndef _LP64
#endif // !LP64
```

Thanks for the review @DamonFool @dholmes-ora @jatin-bhateja.

DamonFool commented:
Looks good to me.
Thanks for your update.

Please make sure the copyright year has been updated (TestSignumIntrinsic.java?) and all the imports are actually needed (import org.openjdk.jmh.annotations.Setup; ?).

mgkwill commented:
> Looks good to me.
> Thanks for your update.
> 
> Please make sure the copyright year has been updated (TestSignumIntrinsic.java?) and all the imports are actually needed (import org.openjdk.jmh.annotations.Setup; ?).

Thanks @DamonFool. Updated with your suggestions.

DamonFool commented on test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```
This is incorrect since it's not a new file.

I suggest this kind of change
```
diff --git a/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java b/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
index 86434d556bb..37aa511dcff 100644
--- a/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
+++ b/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, 2021, Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 2020, BELLSOFT. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
```

mgkwill replied:
Thanks again @DamonFool. Fixed as you suggested.

DamonFool reviewed (approved):
Looks good to me.
Thanks.

mgkwill commented:
Thanks for the review @sviswa7. 

I've updated to `StubRoutines::x86::float_sign_flip()` & `StubRoutines::x86::double_sign_flip()` as I was using when the code was in the src/hotspot/cpu/x86/x86.ad. If I understood your comment correctly this should use 128 bit aligned.

I've added comments as suggested.

sviswa7 commented:
@mgkwill I was wrong. It looks like the CodeEntryAlignment is set to 16 bytes i.e. 128 bits as minimum (globals_x86.hpp). So please continue to use the vector_float_sign_flip and vector_double_sign_flip masks. Xorps and xorpd are vector instructions and we should use the vector masks here.

mgkwill commented:
> @mgkwill I was wrong. It looks like the CodeEntryAlignment is set to 16 bytes i.e. 128 bits as minimum (globals_x86.hpp). So please continue to use the vector_float_sign_flip and vector_double_sign_flip masks. Xorps and xorpd are vector instructions and we should use the vector masks here.

Thanks @sviswa7. I've reverted to `StubRoutines::x86::vector_float_sign_flip(`) `StubRoutines::x86::vector_double_sign_flip()`

sviswa7 commented:
Looks good to me. Thanks.

jatin-bhateja commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -1045,6 +1045,42 @@ void C2_MacroAssembler::evminmax_fp(int opcode, BasicType elem_bt,
+  jcc(Assembler::equal, DONE_LABEL);    // handle special case +0.0/-0.0, if argument is +0.0/-0.0, return argument
+  jcc(Assembler::parity, DONE_LABEL);   // handle special case NaN, if argument NaN, return NaN
+
+  if (opcode == Op_SignumF){
```
Too many if/else, can you kindly club the instruction for SignumF and SignumD in one conditional(if/else) block.

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5785,32 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
+// --------------------------------- Signum ---------------------------
+
+instruct signumF_reg(regF dst, regF zero, regF one, rRegP scratch, rFlagsReg cr) %{
+  predicate(UseSSE>=1);
```
Predicate looks redundant since you already added a check in match_rule_supported

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -5775,6 +5785,32 @@ instruct evminmaxFP_reg_eavx(vec dst, vec a, vec b, vec atmp, vec btmp, kReg ktm
+%}
+
+instruct signumD_reg(regD dst, regD zero, regD one, rRegP scratch, rFlagsReg cr) %{
+  predicate(UseSSE>=2);
```
Same as above

mgkwill commented:
Thanks for your review @jatin-bhateja. I've implemented your suggested changes.

jatin-bhateja commented:
> Thanks for your review @jatin-bhateja. I've implemented your suggested changes.

Thanks Marcus!

sviswa7 commented:
@vnkozlov @neliasso Could you please help review this small patch from Marcus.

neliasso reviewed (approved):
Looks good!

sviswa7 commented:
@neliasso Thanks a lot for the review.

mgkwill commented:
/integrate

sviswa7 commented:
/sponsor
